implement --list-chroots command

xsuchy opened the pull request:
This prints list of chroots available on system. Including users ones.
If the chroot has ["description"] then it is printed beside the chroot
name.

Notes:
- I clear atexit functions during parsing config. Otherwise during
  sys.exit() in the fork the exit function in syslinux plugin is called
  (and it prints error)
- I temporary disable logging before parsing the configs. Otherwise it
  prints "INFO: reading config ..." before reading the config, which
  would clutter the output.

And I still have to edit man page and bash completation.

pep8speaks commented:
Hello @xsuchy! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`mock/py/mock.py`](https://github.com/rpm-software-management/mock/blob/ac886858b28ffceac4baa0bb9cbf314e6666674f/mock/py/mock.py):

> [Line 592:1](https://github.com/rpm-software-management/mock/blob/ac886858b28ffceac4baa0bb9cbf314e6666674f/mock/py/mock.py#L592): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

* In the file [`mock/py/mockbuild/config.py`](https://github.com/rpm-software-management/mock/blob/ac886858b28ffceac4baa0bb9cbf314e6666674f/mock/py/mockbuild/config.py):

> [Line 732:1](https://github.com/rpm-software-management/mock/blob/ac886858b28ffceac4baa0bb9cbf314e6666674f/mock/py/mockbuild/config.py#L732): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 737:80](https://github.com/rpm-software-management/mock/blob/ac886858b28ffceac4baa0bb9cbf314e6666674f/mock/py/mockbuild/config.py#L737): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 739:1](https://github.com/rpm-software-management/mock/blob/ac886858b28ffceac4baa0bb9cbf314e6666674f/mock/py/mockbuild/config.py#L739): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 774:121](https://github.com/rpm-software-management/mock/blob/ac886858b28ffceac4baa0bb9cbf314e6666674f/mock/py/mockbuild/config.py#L774): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (127 > 120 characters)

Conan-Kudo commented on mock-core-configs/etc/mock/alma+epel-8-aarch64.cfg:
```diff
@@ -2,5 +2,6 @@ include('templates/almalinux-8.tpl')
 include('templates/epel-8.tpl')
 
 config_opts['root'] = 'alma+epel-8-aarch64'
+config_opts['description'] = 'Alma Linux 8 + EPEL'
```
`Alma Linux` -> `AlmaLinux`

Conan-Kudo commented:
@xsuchy Maybe we need some kind of cache of the detected names and descriptions?

xsuchy commented:
I have separated the logic of listing the chroot names. To ease the use in future feature --scrub-all.

xsuchy commented:
Updated and ready for review. Especially upper/lower case and spaces in FooLinux in descriptions.
I added a man page. Bash completation works without a change.

I will add caching as suggested by @Conan-Kudo in separate PR.

Conan-Kudo reviewed (approved):
Looks good to me!

praiskup commented on mock/py/mockbuild/config.py:
```diff
@@ -728,6 +730,52 @@ def load_defaults(uidManager, version, pkg_python_dir):
 
+def print_description(config_path, config_filename, uidManager, version, pkg_python_dir):
+    (basename, dirname, basename_without_ext) = parse_config_filename(config_filename)
+    config_opts = load_config(config_path, config_filename, uidManager, version, pkg_python_dir)
```
Can you please catch exceptions from this method?  I can see this on my box:
mockbuild.exception.ConfigError: Could not find included config file: /etc/mock/epel-8-aarch64.cfg

That's because of:
```
$ rpm -qf /etc/mock/epel-8-aarch64-rpmfusion_free.cfg 
mock-rpmfusion-free-36.0-1.fc35.noarch
```

xsuchy commented:
The output now looks  like this:

```
rocky-8-aarch64                    Rocky Linux 8
rocky-8-x86_64                     Rocky Linux 8
Custom configs:

mockbuild.exception.ConfigError: Could not find included config file: /etc/mock/foohkhk

fedora-50-x86_64                   error during parsing the config file
fedora-rawhide-python39            Fedora Rawhide
```
This is best I can do. It is not as easy as I expected. I had to add load_deafaults() to that except branch because without it, python still try to read the include again and again (and I admit that I do not know why) and starts a fork() bomb.

praiskup commented on mock/py/mockbuild/config.py:
```diff
@@ -728,6 +730,56 @@ def load_defaults(uidManager, version, pkg_python_dir):
+    log = logging.getLogger()
+    log.disabled = True
+    # array to save config paths
+    print("Global configs:")
```
I would print the Custom configs first, just because there are a few of them ... and are kind of more "interesting" usually (if something goes wrong with mock config, or so).

xsuchy replied:
The list is long. So I welcome the interesting things towards the end - no need to scroll.

praiskup replied:
I would hate waiting till the end of the command, honestly.  But if the performance will be fixed.

praiskup commented on mock/py/mockbuild/config.py:
```diff
@@ -728,6 +730,56 @@ def load_defaults(uidManager, version, pkg_python_dir):
+    return result
+
+
+def get_user_configs(config_opts, uidManager):
```
Nit:  I would use `*_user_config_files()` naming instead of `*_user_configs`.

xsuchy replied:
I do not mind. Changed.

praiskup commented on mock/py/mockbuild/config.py:
```diff
@@ -728,6 +730,56 @@ def load_defaults(uidManager, version, pkg_python_dir):
 
 
+def print_description(config_path, config_filename, uidManager, version, pkg_python_dir):
+    (basename, dirname, basename_without_ext) = parse_config_filename(config_filename)
```
basename and dirname is unused

xsuchy commented on mock/py/mockbuild/config.py:
```diff
@@ -728,6 +730,56 @@ def load_defaults(uidManager, version, pkg_python_dir):
+    print("Global configs:")
+    for config_filename in sorted(get_global_configs(config_opts)):
+        print_description(config_opts['config_path'], config_filename, uidManager, version, pkg_python_dir)
+    print("Custom configs:")
```
Nod. Changed.

xsuchy commented:
I just pushed here the amended commit with all changes.

praiskup commented:
+1
